langchain/src/sql_database.ts

```diff
@@ -0,0 +1,130 @@
+import { DataSource, DataSourceOptions } from "typeorm";
+import {
+  generateTableInfoFromTables,
+  getTableAndColumnsName,
+  SerializedSqlDatabase,
+  SqlDatabaseDataSourceParams,
+  SqlDatabaseOptionsParams,
+  SqlTable,
+  verifyIgnoreTablesExistInDatabase,
+  verifyIncludeTablesExistInDatabase,
+  verifyListTablesExistInDatabase,
+} from "./util/sql_utils.js";
+
+export class SqlDatabase
+  implements SqlDatabaseOptionsParams, SqlDatabaseDataSourceParams
+{
+  appDataSourceOptions: DataSourceOptions;
+
+  appDataSource: DataSource;
+
+  allTables: Array<SqlTable> = [];
+
+  includesTables: Array<string> = [];
+
+  ignoreTables: Array<string> = [];
+
+  sampleRowsInTableInfo = 3;
+
+  protected constructor(fields: SqlDatabaseDataSourceParams) {
+    this.appDataSource = fields.appDataSource;
+    this.appDataSourceOptions = fields.appDataSource.options;
+    if (fields?.includesTables && fields?.ignoreTables) {
+      throw new Error("Cannot specify both include_tables and ignoreTables");
+    }
+    this.includesTables = fields?.includesTables ?? [];
+    this.ignoreTables = fields?.ignoreTables ?? [];
+    this.sampleRowsInTableInfo =
+      fields?.sampleRowsInTableInfo ?? this.sampleRowsInTableInfo;
+  }
+
+  static async fromDataSourceParams(
```

Comment:
this was added to allow passing in `typeorm.DataSource` directly (made testing easier and some users might want that)